posttowebhooks.go

```diff
@@ -0,0 +1,178 @@
+//}
+//`
+
+//compltedstatus Analysis completed status
+const compltedstatus = "Completed"
+const failedstatus = "Failed"
+
+//Payload payload to post to the webhooks
+type Payload struct {
+	Msg, Link, LinkText string
+	Completed           bool
+}
+
+//Subscription defines user subscriptions to webhooks
+type Subscription struct {
+	id, templatetype, url string
+	topics                []string
+}
+
+//template cache
+var templatesmap map[string]string
+
+//ProcessMessages process the received message for post to webhooks
+func ProcessMessages(d *DBConnection, msgs <-chan amqp.Delivery) {
+	if templatesmap == nil { // call only when template cache is not ready
+		temmap, err := d.getTemplates()
+		if err != nil {
+			Log.Error(err)
+			return
+		}
+		templatesmap = temmap
+	}
+	for delivery := range msgs {
+		Log.Printf("[X] Notification %s", delivery.Body)
+		uid := getUserID(d, delivery.Body)
+		if uid != "" {
+			postToHook(d, uid, delivery.Body)
+		} else {
+			Log.Error("User not found!")
+		}
+	}
+}
+
+//getUserID Get user id for this Notification
+func getUserID(d *DBConnection, msg []byte) string {
+	value, _, _, err := jsonparser.Get(msg, "message", "user")
+	if err != nil {
+		Log.Error(err)
+		return ""
+	}
+	log.Printf("user is %s", string(value))
+	uid, err := d.getUserInfo(string(value) + "@" + config.GetString("user.suffix"))
+	if err != nil {
+		Log.Error(err)
+		return ""
+	}
+	return uid
+}
+
+//post to webhooks
+func postToHook(d *DBConnection, uid string, msg []byte) {
+	subs, err := d.getUserSubscriptions(uid)
+	if err != nil {
+		Log.Error(err)
+		return
+	}
+	Log.Printf("No. of subscriptions found: %d", len(subs))
+	if len(subs) > 0 {
+		for _, v := range subs {
+			if isNotificationInTopic(msg, v.topics) {
+				resp, err := http.Post(v.url, "application/json", preparePayloadFromTemplate(templatesmap[v.templatetype], msg))
+				if err != nil {
+					Log.Printf("Error posting to hook %s", err)
+				}
+				defer resp.Body.Close()
+			}
+		}
+	}
+}
+
+//isNotificationInTopic check if user is subscribed to this notification topic
+func isNotificationInTopic(msg []byte, topics []string) bool {
+	value, _, _, err := jsonparser.Get(msg, "message", "type")
+	if err != nil {
+		Log.Error(err)
+		return false
+	}
+
+	if len(topics) < 1 {
+		return false
+	}
+
+	for _, to := range topics {
+		if string(value) == to {
+			Log.Printf("Subscription topic found: %s", to)
+			return true
+		}
+	}
+	return false
+
+}
+
+//Prepare payload from template
+func preparePayloadFromTemplate(templatetext string, msg []byte) *strings.Reader {
+	var buf1 bytes.Buffer
+	var postbody Payload
+	t := template.Must(template.New("slack").Parse(templatetext))
+	w := io.MultiWriter(&buf1)
+	isCompleted := isAnalysisNotifiction(msg) && isAnalysisCompleted(msg)
+	postbody = Payload{getMessage(msg), config.GetString("de.base") + getResultFolder(msg), "Go to results folder in DE", isCompleted}
```

Comment:
I'm picking nits here, but I usually like to include the field names in structure initializers.